hikari/impl/shard.py

```diff
@@ -896,28 +884,54 @@ def _log_debug_payload(self, payload: str, message: str, *args: typing.Any) -> N
 
         self._logger.debug(message, *args)
 
-    def _serialize_presence_payload(self) -> data_binding.JSONObject:
-        idle_since = self._idle_since if self._idle_since is not undefined.UNDEFINED else None
-        afk = self._is_afk if self._is_afk is not undefined.UNDEFINED else False
-        status = self._status if self._status is not undefined.UNDEFINED else presences.Status.ONLINE
-        activity = self._activity if self._activity is not undefined.UNDEFINED else None
-        presence_pl = data_binding.JSONObjectBuilder()
+    def _serialize_and_store_presence_payload(
+        self,
+        idle_since: undefined.UndefinedNoneOr[datetime.datetime] = undefined.UNDEFINED,
+        afk: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        status: undefined.UndefinedOr[presences.Status] = undefined.UNDEFINED,
+        activity: undefined.UndefinedNoneOr[presences.Activity] = undefined.UNDEFINED,
+    ) -> data_binding.JSONObject:
+        payload = data_binding.JSONObjectBuilder()
+
+        if activity is undefined.UNDEFINED:
+            activity = self._activity
+        else:
+            self._activity = activity
 
-        if activity is not None:
-            presence_pl.put("game", {"name": activity.name, "url": activity.url, "type": activity.type})
+        if status is undefined.UNDEFINED:
+            status = self._status
         else:
-            presence_pl.put("game", None)
+            self._status = status
+
+        if idle_since is undefined.UNDEFINED:
+            idle_since = self._idle_since
+        else:
+            self._idle_since = idle_since
+
+        if afk is undefined.UNDEFINED:
+            afk = self._is_afk
+        else:
+            self._is_afk = afk
+
+        payload.put("since", idle_since, conversion=self._serialize_datetime)
+        payload.put("afk", afk)
+        payload.put("status", status, conversion=lambda s: typing.cast(str, s.value))
+        payload.put("game", activity, conversion=self._serialize_activity)
+        return payload
+
+    @staticmethod
+    def _serialize_datetime(dt: typing.Optional[datetime.datetime]) -> typing.Optional[int]:
+        if dt is None:
+            return None
 
-        presence_pl.put("since", int(idle_since.timestamp() * 1_000) if idle_since is not None else None)
-        presence_pl.put("afk", afk)
+        return int(dt.timestamp() * 1_000)
 
-        # Turns out Discord don't document this properly. I can send "offline"
-        # to the gateway, but it will actually just result in the bot not
-        # changing the status. I have to set it to "invisible" instead to get
-        # this to work...
-        presence_pl.put("status", "invisible" if status is presences.Status.OFFLINE else status)
+    @staticmethod
+    def _serialize_activity(activity: typing.Optional[presences.Activity]) -> data_binding.JSONish:
```

Comment:
Doesn't this return `typing.Optional[data_binding.JSONish]`?

Reply:
No, that wouldn't make sense

Reply:
Code shows a bit under that if activity is none, it will return `None`:
```py
    @staticmethod
    def _serialize_activity(activity: typing.Optional[presences.Activity]) -> data_binding.JSONish:
        if activity is None:
            return None

        return {"name": activity.name, "type": int(activity.type), "url": activity.url}
```

Reply:
Just realized `None` is in that Union, my bad :sweat_smile: